zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/AuthorizationCheckBehavior.java

```diff
@@ -57,6 +57,11 @@ public boolean isAuthorized(final AuthorizationRequest request) {
       return true;
     }
 
+    if (!request.getCommand().hasRequestMetadata()) {
+      // The command is written by Zeebe internally. Internal Zeebe commands are always authorized
+      return true;
+    }
```

Comment:
@korthout I'm 99% sure this is safe. I need to identify whether a command is written by Zeebe, or by a user. Checking the request mestadata seems safe to me. Do you know any reasons cases where this wouldn't be true?

Reply:
At this time, all commands that have request metadata are user commands. However, we did once discuss the idea of responding to inter-partition commands rather than sending an ACK command. There may be reasons in the future to change this.

We could consider using the `sourceRecordPosition` for this. Only commands that are the result of processing have a source record position set. But there are some exceptions like commands written by checkers that in addition to user commands also have no source record position. So, it may be hard to use.

Another option to consider is the RECORDS_BY_TYPE of [CommandApiRequestReader](https://github.com/camunda/camunda/blob/13d83cfd8b5d4fd2f5106ac1d54b963a8b031cee/zeebe/broker/src/main/java/io/camunda/zeebe/broker/transport/commandapi/CommandApiRequestReader.java#L53). This is filled with command types that can be sent through the broker's command api.

Reply:
> At this time, all commands that have request metadata are user commands. However, we did once discuss the idea of responding to inter-partition commands rather than sending an ACK command. There may be reasons in the future to change this.

I think that should be fine because we don't verify authorizations on distributed commands. I don't see issues there.

> We could consider using the sourceRecordPosition for this. Only commands that are the result of processing have a source record position set. But there are some exceptions like commands written by checkers that in addition to user commands also have no source record position. So, it may be hard to use.

Yes what you described is exactly why it's not possible to use as it is now.

> Another option to consider is the RECORDS_BY_TYPE of [CommandApiRequestReader](https://github.com/camunda/camunda/blob/13d83cfd8b5d4fd2f5106ac1d54b963a8b031cee/zeebe/broker/src/main/java/io/camunda/zeebe/broker/transport/commandapi/CommandApiRequestReader.java#L53). This is filled with command types that can be sent through the broker's command api.

I don't see how this would work. There we only store the value types, not the intents. So while we may say you need to check messages, we only want to do that for the `PUBLISH` intent, as the `EXPIRE` is internal. This becomes more complex with signals, because even if you say you need to check broadcasts, that's not entirely true as signals are broadcasted internally as well.

----------

My conclusion is I think the current approach is fine, even if we would change distirbution to this mechanism. If we later reach a state where we cannot use the request metadata here we can always change it.

Reply:
Okay. One last idea to consider is a strict mapping of supported intents as user commands. But I'm not sure yet what that should look like. It's something you could consider exploring still.

Reply:
I feel like a mapping still doesn't solve the problem where we have a single intent used both internally and externally such as signal broadcasting. You'd always have to check for something else in combination.